src/app/CommandHandler.h

```diff
@@ -175,13 +175,40 @@ class CommandHandler
     template <typename CommandData>
     CHIP_ERROR AddResponseData(const ConcreteCommandPath & aRequestCommandPath, const CommandData & aData)
     {
-        ConcreteCommandPath path = { aRequestCommandPath.mEndpointId, aRequestCommandPath.mClusterId, CommandData::GetCommandId() };
-        ReturnErrorOnFailure(PrepareCommand(path, false));
-        TLV::TLVWriter * writer = GetCommandDataIBTLVWriter();
-        VerifyOrReturnError(writer != nullptr, CHIP_ERROR_INCORRECT_STATE);
-        ReturnErrorOnFailure(DataModel::Encode(*writer, TLV::ContextTag(to_underlying(CommandDataIB::Tag::kData)), aData));
+        // TryAddResponseData will ensure we are in the correct state when calling AddResponseData.
+        CHIP_ERROR err = TryAddResponseData(aRequestCommandPath, aData);
+        if (err != CHIP_NO_ERROR)
+        {
+            // We have verified the state above, so this call must succeed since we must be in one of the states required by
+            // RollbackResponse.
+            RollbackResponse();
+        }
+        return err;
+    }
 
-        return FinishCommand(/* aEndDataStruct = */ false);
+    /**
+     * API for adding a response.  This will try to encode a data response (response command), and if that fails will encode a a
+     * Protocols::InteractionModel::Status::Failure status response instead.
+     *
+     * The template parameter T is generally expected to be a ClusterName::Commands::CommandName::Type struct, but any object that
+     * can be encoded using the DataModel::Encode machinery and exposes the right command id will work.
+     *
+     * Since the function will call AddStatus when failed to encode the data, it cannot send any response when it failed to encode a
+     * status code since another AddStatus call will also fail. Users can check the log when thry failed to receive a response.
```

Comment:
```suggestion
     * Since the function will call AddStatus when it fails to encode the data, it cannot send any response when it fails to encode a
     * status code since another AddStatus call will also fail. The error from AddStatus will just be logged.
```